### ISSUE TITLE
Setting a maximum number of trials per study

### ISSUE THREAD

**Issue by jtourille** [optuna/optuna#1883]
I am moving from hyperopt to optuna and it has been a nice transition so far. This is a very nice piece of code!
One thing that I liked in hyperopt and didn't find in optuna is the ability to specify a maximum number of trials for one study when doing distributed optimization.

## Motivation

In optuna, the argument `n_trials` is the number of trials for one worker. So if I specify 4 trials and launch 4 workers (i.e. 4 executions of my script), I will get 16 trials in my study.

However, some hyperparameter combinations take longer to be tested than others. At some point, I could have 3 out my 4 workers that are done with their trials and the last one with 3 remaining trials to perform. This is a situation where this feature could be useful.

Setting a maximum number of trials for a study could allow to better optimize the time spent for one study. The more workers you add, the quicker the study will be performed. 

## Description

I didn't get too deep into reading the code, so I don't have a clear plan on how this could be done :-(
One high-level solution could be to specify a maximum number of trials during study creation. The `optimize` method could be called without a `n_trials` value and stop performing trials once the database contains enough trials.

**Comment by jtourille** [comment]
I guess I can use the `stop()` function of the study object as mentioned in the documentation.

```python
import optuna

def objective(trial):
    if trial.number == 4:
        study.stop()
    x = trial.suggest_uniform("x", 0, 10)
    return x ** 2

study = optuna.create_study()
study.optimize(objective, n_trials=10)
assert len(study.trials) == 5
```

**Comment by jtourille** [comment]
A more complete solution, based on the example found in the documentation. In this example, the study stops after 100 trials.

```python
import optuna

def objective(trial, n_trials):
    if trial.number == n_trials - 1:
        study.stop()
    x = trial.suggest_uniform("x", 0, 10)
    return x ** 2

_ = optuna.create_study(
    study_name="test",
    storage="sqlite:///database.sqlite",
)

study = optuna.load_study(study_name="test", storage="sqlite:///database.sqlite")
study.optimize(lambda trial: objective(trial, 100), n_trials=500)
assert len(study.trials) == 100
```
I guess I can live with that :-) This is a toy example where the value of the `n_trials` argument of the `optimize()` method is not similar to the value of the `n_trials` argument of the objective function. In practice, they should be the same.

**Comment by jtourille** [comment]
Better solution to avoid using `optuna.TrialPruned()`

```python
import optuna

class MaximumNumberOfTrialsException(Exception):
    pass

def objective(trial, n_trials):
    if trial.number >= n_trials:
        study.stop()
        raise MaximumNumberOfTrialsException
    x = trial.suggest_uniform("x", 0, 10)
    return x ** 2

_ = optuna.create_study(
    study_name="test",
    storage="sqlite:///database.sqlite",
)

study = optuna.load_study(study_name="test", storage="sqlite:///database.sqlite")
study.optimize(lambda trial: objective(trial, 100), n_trials=500, catch=(MaximumNumberOfTrialsException,))
trials = study.trials_dataframe()
assert len(trials[trials.state == "COMPLETE"]) == 100
```

**Comment by toshihikoyanase** [comment]
Hi @jtourille ,

> The optimize method could be called without a n_trials value and stop performing trials once the database contains enough trials.

> I guess I can use the stop() method of the study object as mentioned in the documentation.

That's a good idea! I guess the `callbacks` argument of `Study.optimize` might simplify the code.

```python
import optuna

def max_trial_callback(study, trial):
    n_complete = len([t for t in study.trials if t.state == optuna.trial.TrialState.COMPLETE])
    if n_complete >= 100:
        study.stop()

def objective(trial):
    x = trial.suggest_uniform("x", 0, 10)
    return x ** 2

study = optuna.create_study()
study.optimize(objective, callbacks=[max_trial_callback])
trials = study.trials_dataframe()
assert len(trials[trials.state == "COMPLETE"]) == 100
```

Recently, we have a similar issue (#1865), and I think it might be helpful if we provide an example or FAQ page about this topic. What do you think?

**Comment by jtourille** [comment]
Hi @toshihikoyanase,

I agree, your solution is cleaner. One slight improvement could be to check also for running trials to avoid the situation where you get `n` additional trials corresponding to the `n` workers. 

```python
import optuna

def max_trial_callback(study, trial):
    n_complete = len([t for t in study.trials if t.state == optuna.trial.TrialState.COMPLETE or t.state == optuna.trial.TrialState.RUNNING])
    if n_complete >= 100:
        study.stop()

def objective(trial):
    x = trial.suggest_uniform("x", 0, 10)
    return x ** 2

study = optuna.create_study()
study.optimize(objective, callbacks=[max_trial_callback])
trials = study.trials_dataframe()
assert len(trials[trials.state == "COMPLETE"]) == 100
```

> Recently, we have a similar issue (#1865), and I think it might be helpful if we provide an example or FAQ page about this topic. What do you think?

This would be nice as I think this question will arise frequently.

**Comment by toshihikoyanase** [comment]
> check also for running trials to avoid the situation where you get `n` additional trials corresponding to the `n` workers.

That makes sense. Thank you for sharing your hack!

> This would be nice as I think this question will arise frequently.

Thanks. I think your example code is also suitable for `optuna/examples`. If you are interested in a contribution to Optuna, your PR is highly welcome. Otherwise, let me add the `contribution-welcome` label for Hacktoberfest participants.

**Comment by aitikgupta** [comment]
I think this is pretty straightforward, would it make sense to:
- have this addition to [an existing example](https://github.com/optuna/optuna/blob/master/examples/quadratic_simple.py)?
- Or a new `custom_callbacks.py`, which explains one or two types of usecases, such as `max_trial_callback` discussed above.

I'll create a PR if contributions are still welcome. 👍🏼

**Comment by RandomDefaultUser** [comment]
I was looking for a way to ensure that my distributed optuna runs adheres to a set number of trials. I like @jtourille idea but I just wanted to point out one potential problem: In cases were the optimization has to be paused/is interrupted (e.g. due to runtime restrictions on HPC infrastructure) this solution will not work. As pointed out by #1038 and #913, interruption during distributed runs will result in "Zombie" trials that appear as `RUNNING` in the sutdy, therefore  `len([t for t in study.trials if t.state == optuna.trial.TrialState.COMPLETE or t.state == optuna.trial.TrialState.RUNNING])` will not reflect the _actual_ number of completed and running trials. I think one could do `len([t for t in study.trials if t.state == optuna.trial.TrialState.COMPLETE])` instead, and prune trials that are beyond the desired number of trials, at the cost of having to run `n_workers` trials. I will look into this.

**Comment by hvy** [comment]
Since this is a reoccurring pattern, we've merged an experimental callback class which will be available in 2.8.0 (or the current `master` branch). You can by the way specify which trial states to include the count as well, i.e. exclude running trials https://github.com/optuna/optuna/pull/2612

(You are correct about the performance issue. Whether this is problematic or not in practice depends on your objective function, the sampler, etc. In favor of the `RDBStorage` I'm wondering if we shouldn't introduce a method to the `BaseStorage` to return the trial count (likely with a state filtering parameter).)

**Comment by RandomDefaultUser** [comment]
I saw the callback function, but I still think it will not solve the issue with the "Zombie" trials. While it allows to exclude running trials, this exclusion would then also apply to the actually still running trials on other workers (as mentioned @jtourille) and lead to up to `n_workers-1` more trials in the finished study compared to what was requested. I tried it out yesterday, using 4 workers and requesting 1000 trials. I interrupted each worker once and restarted it. In the end, I wound up with 1003 completed trials. 
Please don't get me wrong, the callback is very useful and I will be using this functionality. But in order to ensure that only the precise number of trials is included in the study I believe one would have to be able to delete these "Zombie" trials after interruption. I realize from reading through #913 and #1038 that this functionality is not trivial to implement. I just wanted to point out that I do not see another way to ensure that the maximum number of trials is respected for distributed optimization.

**Comment by hvy** [comment]
It's not a waterproof solution but in order to detect zombie trials from interrupted processes, you could try the heartbeat feature https://optuna.readthedocs.io/en/latest/reference/generated/optuna.storages.RDBStorage.html?highlight=heartbeat. This'll make sure that trials that been running for a certain amount of time without any heartbeat will automatically be failed.

**Comment by RandomDefaultUser** [comment]
Thanks @hvy , that sounds like something I was looking for! I will try this out.

**Comment by RandomDefaultUser** [comment]
I tried it out and it gives the desired behavior in distributed optimization . I did observe the same behavior as mentioned in #2685 , but since this is fixed by #2686 and the just released 2.8.0, this should not be a problem.

**Comment by Crissman** [comment]
I believe the initial request by @jtourille has been resolved, so closing this issue.

If there is a new issue, please open.

Thanks!
